In [3]:
import pandas as pd

In [4]:
df = pd.read_csv('data/wine_cleared.csv')

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129971 entries, 0 to 129970
Data columns (total 13 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   Unnamed: 0             129971 non-null  int64  
 1   country                129908 non-null  object 
 2   description            129971 non-null  object 
 3   designation            129971 non-null  object 
 4   points                 129971 non-null  int64  
 5   price                  129971 non-null  float64
 6   province               129908 non-null  object 
 7   region_1               129971 non-null  object 
 8   taster_name            129971 non-null  object 
 9   taster_twitter_handle  129971 non-null  object 
 10  title                  129971 non-null  object 
 11  variety                129970 non-null  object 
 12  winery                 129971 non-null  object 
dtypes: float64(1), int64(2), object(10)
memory usage: 12.9+ MB


In [6]:
df.drop(columns=['Unnamed: 0'], inplace=True)

In [7]:
df.head()

country                                        description  \
0     Italy  Aromas include tropical fruit, broom, brimston...   
1  Portugal  This is ripe and fruity, a wine that is smooth...   
2        US  Tart and snappy, the flavors of lime flesh and...   
3        US  Pineapple rind, lemon pith and orange blossom ...   
4        US  Much like the regular bottling from 2012, this...   

                          designation  points      price           province  \
0                        Vulkà Bianco      87  35.363389  Sicily & Sardinia   
1                            Avidagos      87  15.000000              Douro   
2                             unknown      87  14.000000             Oregon   
3                Reserve Late Harvest      87  13.000000           Michigan   
4  Vintner's Reserve Wild Child Block      87  65.000000             Oregon   

              region_1         taster_name taster_twitter_handle  \
0                 Etna       Kerin O’Keefe          @kerinokeefe   
1              unknown          Roger Voss            @vossroger   
2    Willamette Valley        Paul Gregutt           @paulgwine    
3  Lake Michigan Shore  Alexander Peartree               unknown   
4    Willamette Valley        Paul Gregutt           @paulgwine    

                                               title         variety  \
0                  Nicosia 2013 Vulkà Bianco  (Etna)     White Blend   
1      Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2      Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3  St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4  Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   

                winery  
0              Nicosia  
1  Quinta dos Avidagos  
2            Rainstorm  
3           St. Julian  
4         Sweet Cheeks

In [8]:
df['price_round'] = df['price'].round().astype('int')

In [9]:
df['year'] = df['title'].str.findall('\d{4}').str.get(0).astype('int', copy=False, errors='ignore')

In [10]:
df.head()

country                                        description  \
0     Italy  Aromas include tropical fruit, broom, brimston...   
1  Portugal  This is ripe and fruity, a wine that is smooth...   
2        US  Tart and snappy, the flavors of lime flesh and...   
3        US  Pineapple rind, lemon pith and orange blossom ...   
4        US  Much like the regular bottling from 2012, this...   

                          designation  points      price           province  \
0                        Vulkà Bianco      87  35.363389  Sicily & Sardinia   
1                            Avidagos      87  15.000000              Douro   
2                             unknown      87  14.000000             Oregon   
3                Reserve Late Harvest      87  13.000000           Michigan   
4  Vintner's Reserve Wild Child Block      87  65.000000             Oregon   

              region_1         taster_name taster_twitter_handle  \
0                 Etna       Kerin O’Keefe          @kerinokeefe   
1              unknown          Roger Voss            @vossroger   
2    Willamette Valley        Paul Gregutt           @paulgwine    
3  Lake Michigan Shore  Alexander Peartree               unknown   
4    Willamette Valley        Paul Gregutt           @paulgwine    

                                               title         variety  \
0                  Nicosia 2013 Vulkà Bianco  (Etna)     White Blend   
1      Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2      Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3  St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4  Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   

                winery  price_round  year  
0              Nicosia           35  2013  
1  Quinta dos Avidagos           15  2011  
2            Rainstorm           14  2013  
3           St. Julian           13  2013  
4         Sweet Cheeks           65  2012

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129971 entries, 0 to 129970
Data columns (total 14 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   country                129908 non-null  object 
 1   description            129971 non-null  object 
 2   designation            129971 non-null  object 
 3   points                 129971 non-null  int64  
 4   price                  129971 non-null  float64
 5   province               129908 non-null  object 
 6   region_1               129971 non-null  object 
 7   taster_name            129971 non-null  object 
 8   taster_twitter_handle  129971 non-null  object 
 9   title                  129971 non-null  object 
 10  variety                129970 non-null  object 
 11  winery                 129971 non-null  object 
 12  price_round            129971 non-null  int32  
 13  year                   125362 non-null  object 
dtypes: float64(1), int32(1), int64(1), o

In [12]:
df['is_usa'] = df['country'].apply(lambda x: 1 if x == 'US' else 0)

###  Задание 2.1 ###

Выбрать из списка две самых популярных (помимо США) страны, производящих вино.

***Ответ - France, Italy***

In [13]:
df['country'].value_counts()

US                        54504
France                    22093
Italy                     19540
Spain                      6645
Portugal                   5691
Chile                      4472
Argentina                  3800
Austria                    3345
Australia                  2329
Germany                    2165
New Zealand                1419
South Africa               1401
Israel                      505
Greece                      466
Canada                      257
Hungary                     146
Bulgaria                    141
Romania                     120
Uruguay                     109
Turkey                       90
Slovenia                     87
Georgia                      86
England                      74
Croatia                      73
Mexico                       70
Moldova                      59
Brazil                       52
Lebanon                      35
Morocco                      28
Peru                         16
Ukraine                      14
Serbia  

###  Задание 2.2 ###

Создать бинарные признаки `is_france`, `is_italy` наподобие признака `is_usa`.

В ответ вписать результат выполнения кода `data['is_france'].sum()` и `data['is_italy'].sum()`

***Ответ - 22093, 19540***

In [14]:
df['is_france'] = df['country'].apply(lambda x: 1 if x == 'France' else 0)
df['is_italy'] = df['country'].apply(lambda x: 1 if x == 'Italy' else 0)
df['is_france'].sum()

22093

In [15]:
df['is_italy'].sum()

19540

###  Задание 2.3 ###

Создать новый бинарный признак `old_wine`, где значение **1** будет обозначать, что вино старше 2010 года.

В ответ вписать результат выполнения кода `data['old_wine'].sum()`

***Ответ - 39781***

In [16]:
df['old_wine'] = df['year'].apply(lambda x: 1 if pd.to_numeric(x) < 2010 else 0)

In [17]:
df['old_wine'].sum()

39781

###  Задание 2.7 (Самопроверка) ###

Создать новый признак `locality` из признака `title`, который будет обозначать название долины/местности производства вина.

Например, в названии вина **Rainstorm 2013 Pinot Gris (Willamette Valley)** `locality` будет **Willamette Valley**. В названии **Tandem 2011 Ars In Vitro Tempranillo-Merlot (Navarra)**&nbsp;&mdash; **Navarra**.

In [18]:
df['locality'] = df['title'].str.findall(r'\(.+\)').str.get(0).str.replace('\(|\)', '', regex=True)

In [19]:
df.head(20)

country                                        description  \
0       Italy  Aromas include tropical fruit, broom, brimston...   
1    Portugal  This is ripe and fruity, a wine that is smooth...   
2          US  Tart and snappy, the flavors of lime flesh and...   
3          US  Pineapple rind, lemon pith and orange blossom ...   
4          US  Much like the regular bottling from 2012, this...   
5       Spain  Blackberry and raspberry aromas show a typical...   
6       Italy  Here's a bright, informal red that opens with ...   
7      France  This dry and restrained wine offers spice in p...   
8     Germany  Savory dried thyme notes accent sunnier flavor...   
9      France  This has great depth of flavor with its fresh ...   
10         US  Soft, supple plum envelopes an oaky structure ...   
11     France  This is a dry wine, very spicy, with a tight, ...   
12         US  Slightly reduced, this wine offers a chalky, t...   
13      Italy  This is dominated by oak and oak-driven aromas...   
14         US  Building on 150 years and six generations of w...   
15    Germany  Zesty orange peels and apple notes abound in t...   
16  Argentina  Baked plum, molasses, balsamic vinegar and che...   
17  Argentina  Raw black-cherry aromas are direct and simple ...   
18      Spain  Desiccated blackberry, leather, charred wood a...   
19         US  Red fruit aromas pervade on the nose, with cig...   

                                          designation  points      price  \
0                                        Vulkà Bianco      87  35.363389   
1                                            Avidagos      87  15.000000   
2                                             unknown      87  14.000000   
3                                Reserve Late Harvest      87  13.000000   
4                  Vintner's Reserve Wild Child Block      87  65.000000   
5                                        Ars In Vitro      87  15.000000   
6                                             Belsito      87  16.000000   
7                                             unknown      87  24.000000   
8                                               Shine      87  12.000000   
9                                         Les Natures      87  27.000000   
10                                     Mountain Cuvée      87  19.000000   
11                                            unknown      87  30.000000   
12                                            unknown      87  34.000000   
13                                              Rosso      87  35.363389   
14                                            unknown      87  12.000000   
15                                              Devon      87  24.000000   
16                                              Felix      87  30.000000   
17                                Winemaker Selection      87  13.000000   
18  Vendimia Seleccionada Finca Valdelayegua Singl...      87  28.000000   
19                                            unknown      87  32.000000   

             province             region_1         taster_name  \
0   Sicily & Sardinia                 Etna       Kerin O’Keefe   
1               Douro              unknown          Roger Voss   
2              Oregon    Willamette Valley        Paul Gregutt   
3            Michigan  Lake Michigan Shore  Alexander Peartree   
4              Oregon    Willamette Valley        Paul Gregutt   
5      Northern Spain              Navarra   Michael Schachner   
6   Sicily & Sardinia             Vittoria       Kerin O’Keefe   
7              Alsace               Alsace          Roger Voss   
8         Rheinhessen              unknown  Anna Lee C. Iijima   
9              Alsace               Alsace          Roger Voss   
10         California          Napa Valley      Virginie Boone   
11             Alsace               Alsace          Roger Voss   
12         California     Alexander Valley      Virginie Boone   
13  Sicily & Sardinia                 Etna       Kerin O

# 3. Создание признаков. Внешние источники данных #

In [20]:
country_population = pd.read_csv('data/country_population.csv', sep=';')
display(country_population)

country     population
0               China  1,411,778,724
1               India  1,386,584,581
2                  US    333,022,386
3           Indonesia    271,350,000
4            Pakistan    225,200,000
..                ...            ...
236              Niue          1,549
237           Tokelau          1,501
238      Vatican City            825
239     Cocos Islands            573
240  Pitcairn Islands             40

[241 rows x 2 columns]

###  Задание 3.1 ###

Каково население Италии согласно датасету `country_population`? Ответ без пробелов и запятых.

***Ответ - 59097904***

In [21]:
display(country_population[country_population.country == 'Italy'].population)

24    59,097,904
Name: population, dtype: object

In [22]:
df.join(country_population.set_index('country'), on='country')

country                                        description  \
0          Italy  Aromas include tropical fruit, broom, brimston...   
1       Portugal  This is ripe and fruity, a wine that is smooth...   
2             US  Tart and snappy, the flavors of lime flesh and...   
3             US  Pineapple rind, lemon pith and orange blossom ...   
4             US  Much like the regular bottling from 2012, this...   
...          ...                                                ...   
129966   Germany  Notes of honeysuckle and cantaloupe sweeten th...   
129967        US  Citation is given as much as a decade of bottl...   
129968    France  Well-drained gravel soil gives this wine its c...   
129969    France  A dry style of Pinot Gris, this is crisp with ...   
129970    France  Big, rich and off-dry, this is powered by inte...   

                                   designation  points      price  \
0                                 Vulkà Bianco      87  35.363389   
1                                     Avidagos      87  15.000000   
2                                      unknown      87  14.000000   
3                         Reserve Late Harvest      87  13.000000   
4           Vintner's Reserve Wild Child Block      87  65.000000   
...                                        ...     ...        ...   
129966  Brauneberger Juffer-Sonnenuhr Spätlese      90  28.000000   
129967                                 unknown      90  75.000000   
129968                                   Kritt      90  30.000000   
129969                                 unknown      90  32.000000   
129970           Lieu-dit Harth Cuvée Caroline      90  21.000000   

                 province             region_1         taster_name  \
0       Sicily & Sardinia                 Etna       Kerin O’Keefe   
1                   Douro              unknown          Roger Voss   
2                  Oregon    Willamette Valley        Paul Gregutt   
3                Michigan  Lake Michigan Shore  Alexander Peartree   
4                  Oregon    Willamette Valley        Paul Gregutt   
...                   ...                  ...                 ...   
129966              Mosel              unknown  Anna Lee C. Iijima   
129967             Oregon               Oregon        Paul Gregutt   
129968             Alsace               Alsace          Roger Voss   
129969             Alsace               Alsace          Roger Voss   
129970             Alsace               Alsace          Roger Voss   

       taster_twitter_handle  \
0               @kerinokeefe   
1                 @vossroger   
2                @paulgwine    
3                    unknown   
4                @paulgwine    
...                      ...   
129966               unknown   
129967           @paulgwine    
129968            @vossroger   
129969            @vossroger   
129970            @vossroger   

                                                    title         variety  \
0                       Nicosia 2013 Vulkà Bianco  (Etna)     White Blend   
1           Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2           Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3       St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4       Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   
...                                                   ...             ...   
129966  Dr. H. Thanisch (Erben Müller-Burggraef) 2013 ...        Riesling   
129967                  Citation 2004 Pinot Noir (Oregon)      Pinot Noir   
129968  Domaine Gresser 2013 Kritt Gewurztraminer (Als...  Gewürztraminer   
129969      Domaine Marcel Deiss 2012 Pinot Gris (Alsace)      Pinot Gris   
129970  Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...  Gewürztraminer   

                                          winery  price_round  year  is_usa  \
0                                        Nicosia           35  2013       0   
1                 

In [23]:
country_area = pd.read_csv('data/country_area.csv', sep=';')
display(country_area)

country         area
0          Russia  17075400.00
1          Canada   9984670.00
2           China   9596960.00
3              US   9372610.00
4          Brazil   8547000.00
..            ...          ...
188    San Marino        61.00
189        Tuvalu        26.00
190         Nauru        21.00
191        Monaco         2.00
192  Vatican City         0.44

[193 rows x 2 columns]

###  Задание 3.2 ###

Создать новый признак `area_country`&nbsp;&mdash; площадь страны, аналогичный признаку `country_population`.

Какая площадь страны у вина под названием **'Gård 2014 Grand Klasse Reserve Lawrence Vineyards Viognier (Columbia Valley (WA))'**?

Ответ без пробелов, округлить до целых.

***Ответ - 9372610***

In [24]:
df2 = df.join(country_area.set_index('country'), on='country')

In [25]:
df2[df2['title'].str.contains('Reserve Lawrence Vineyards Viognier')].area

94    9372610.0
Name: area, dtype: float64

# 4. Создание признаков. Работа с форматом &laquo;дата-время&raquo; #

In [26]:
calls_list = [
    [460, '2013-12-17 04:55:39', '2013-12-17 04:55:44', '2013-12-17 04:55:45'],
    [12, '2013-12-16 20:03:20', '2013-12-16 20:03:22', '2013-12-16 20:07:13'],
    [56, '2013-12-16 20:03:20', '2013-12-16 20:03:20', '2013-12-16 20:05:04'],
    [980, '2013-12-16 20:03:20','2013-12-16 20:03:27', '2013-12-16 20:03:29'],
    [396, '2013-12-16 20:08:27', '2013-12-16 20:08:28','2013-12-16 20:12:03'],
    [449, '2013-12-16 20:03:20', '2013-12-16 20:03:25','2013-12-16 20:05:00'],
    [397, '2013-12-16 20:08:25', '2013-12-16 20:08:27', '2013-12-16 20:09:59'],
    [398, '2013-12-16 20:01:23', '2013-12-16 20:01:23', '2013-12-16 20:04:58'],
    [452, '2013-12-16 20:03:20', '2013-12-16 20:03:21','2013-12-16 20:04:55'],
    [440, '2013-12-16 20:03:20', '2013-12-16 20:04:26', '2013-12-16 20:04:32']
]
calls = pd.DataFrame(calls_list, columns=['client_id', 'agent_date', 'created_at', 'end_date'])
calls['agent_date'] = pd.to_datetime(calls['agent_date'])
calls['created_at'] = pd.to_datetime(calls['created_at'])
calls['end_date'] = pd.to_datetime(calls['end_date'])
display(calls)

client_id          agent_date          created_at            end_date
0        460 2013-12-17 04:55:39 2013-12-17 04:55:44 2013-12-17 04:55:45
1         12 2013-12-16 20:03:20 2013-12-16 20:03:22 2013-12-16 20:07:13
2         56 2013-12-16 20:03:20 2013-12-16 20:03:20 2013-12-16 20:05:04
3        980 2013-12-16 20:03:20 2013-12-16 20:03:27 2013-12-16 20:03:29
4        396 2013-12-16 20:08:27 2013-12-16 20:08:28 2013-12-16 20:12:03
5        449 2013-12-16 20:03:20 2013-12-16 20:03:25 2013-12-16 20:05:00
6        397 2013-12-16 20:08:25 2013-12-16 20:08:27 2013-12-16 20:09:59
7        398 2013-12-16 20:01:23 2013-12-16 20:01:23 2013-12-16 20:04:58
8        452 2013-12-16 20:03:20 2013-12-16 20:03:21 2013-12-16 20:04:55
9        440 2013-12-16 20:03:20 2013-12-16 20:04:26 2013-12-16 20:04:32

In [27]:
calls.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   client_id   10 non-null     int64         
 1   agent_date  10 non-null     datetime64[ns]
 2   created_at  10 non-null     datetime64[ns]
 3   end_date    10 non-null     datetime64[ns]
dtypes: datetime64[ns](3), int64(1)
memory usage: 448.0 bytes


In [28]:
calls['duration'] = (calls['end_date'] - calls['created_at']).dt.seconds
display(calls)

client_id          agent_date          created_at            end_date  \
0        460 2013-12-17 04:55:39 2013-12-17 04:55:44 2013-12-17 04:55:45   
1         12 2013-12-16 20:03:20 2013-12-16 20:03:22 2013-12-16 20:07:13   
2         56 2013-12-16 20:03:20 2013-12-16 20:03:20 2013-12-16 20:05:04   
3        980 2013-12-16 20:03:20 2013-12-16 20:03:27 2013-12-16 20:03:29   
4        396 2013-12-16 20:08:27 2013-12-16 20:08:28 2013-12-16 20:12:03   
5        449 2013-12-16 20:03:20 2013-12-16 20:03:25 2013-12-16 20:05:00   
6        397 2013-12-16 20:08:25 2013-12-16 20:08:27 2013-12-16 20:09:59   
7        398 2013-12-16 20:01:23 2013-12-16 20:01:23 2013-12-16 20:04:58   
8        452 2013-12-16 20:03:20 2013-12-16 20:03:21 2013-12-16 20:04:55   
9        440 2013-12-16 20:03:20 2013-12-16 20:04:26 2013-12-16 20:04:32   

   duration  
0         1  
1       231  
2       104  
3         2  
4       215  
5        95  
6        92  
7       215  
8        94  
9         6

###  Задание 4.1 ###

Подсчитать, сколько секунд тратят сотрудники компании на дозвон клиенту. Результат записать в новый признак `time_connection`.

В ответ записать результат выполнения следующего кода: `calls['time_connection'].sum()`.

***Ответ - 89***

In [29]:
calls['time_connection'] = (calls['created_at'] - calls['agent_date']).dt.seconds
calls['time_connection'].sum()

89

###  Задание 4.2 ###

Создать новый признак `is_connection`&nbsp;&mdash; факт соединения с клиентом. Признак будет равен 1 в случае, если разговор состоялся и продлился больше 10 секунд, иначе&nbsp;&mdash; 0.

В ответ записать результат выполнения следующего кода: `calls['is_connection'].sum()`.

***Ответ - 7***

In [30]:
calls['is_connection'] = calls['duration'].apply(lambda x: 1 if x > 10 else 0)
calls['is_connection'].sum()

7

###  Задание 4.3 ###

Создать признак `time_diff`&nbsp;&mdash; разницу в секундах между началом звонка(не разговора) и его окончанием.

В ответ записать результат выполнения следующего кода: `calls['time_diff'].sum()`.

***Ответ - 1144***

In [31]:
calls['time_diff'] = (calls['end_date'] - calls['agent_date']).dt.seconds
calls['time_diff'].sum()

1144

In [32]:
display(calls)

client_id          agent_date          created_at            end_date  \
0        460 2013-12-17 04:55:39 2013-12-17 04:55:44 2013-12-17 04:55:45   
1         12 2013-12-16 20:03:20 2013-12-16 20:03:22 2013-12-16 20:07:13   
2         56 2013-12-16 20:03:20 2013-12-16 20:03:20 2013-12-16 20:05:04   
3        980 2013-12-16 20:03:20 2013-12-16 20:03:27 2013-12-16 20:03:29   
4        396 2013-12-16 20:08:27 2013-12-16 20:08:28 2013-12-16 20:12:03   
5        449 2013-12-16 20:03:20 2013-12-16 20:03:25 2013-12-16 20:05:00   
6        397 2013-12-16 20:08:25 2013-12-16 20:08:27 2013-12-16 20:09:59   
7        398 2013-12-16 20:01:23 2013-12-16 20:01:23 2013-12-16 20:04:58   
8        452 2013-12-16 20:03:20 2013-12-16 20:03:21 2013-12-16 20:04:55   
9        440 2013-12-16 20:03:20 2013-12-16 20:04:26 2013-12-16 20:04:32   

   duration  time_connection  is_connection  time_diff  
0         1                5              0          6  
1       231                2              1        233  
2       104                0              1        104  
3         2                7              0          9  
4       215                1              1        216  
5        95                5              1        100  
6        92                2              1         94  
7       215                0              1        215  
8        94                1              1         95  
9         6               66              0         72

In [33]:
calls.drop(columns=['agent_date', 'created_at', 'end_date'], inplace=True)
display(calls)

client_id  duration  time_connection  is_connection  time_diff
0        460         1                5              0          6
1         12       231                2              1        233
2         56       104                0              1        104
3        980         2                7              0          9
4        396       215                1              1        216
5        449        95                5              1        100
6        397        92                2              1         94
7        398       215                0              1        215
8        452        94                1              1         95
9        440         6               66              0         72

###  Задание 4.5 ###

Создать признак количество дней с момента произведения вина&nbsp;&mdash; `years_diff` для датасета винных обзоров. За дату отсчёта взять **12 января 2022 года**. В ответ вписать максимальное количество дней с момента произведения вина.

Ответ округлить до целого числа.

***Ответ -***

In [34]:
# dfy = df[['title', 'year']].dropna(axis=0)
dfy = df[['title', 'year']]
# dfy['year'].value_counts()
# dfy.sort_values(by='year', ignore_index=True)
# dfy[dfy['year'] > '1970'].value_counts()
# dfy['prod_year'] = dfy['year'].apply(lambda x: str(x) + '-01-01')
# dfy['prod_day'] = pd.to_datetime(dfy['prod_year'])
# dfy.info()
dfy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129971 entries, 0 to 129970
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   title   129971 non-null  object
 1   year    125362 non-null  object
dtypes: object(2)
memory usage: 2.0+ MB


# 5. Кодирование признаков. Методы #

In [35]:
##### pip install category_encoders
import category_encoders as ce

In [36]:
clothing_list = [
    ['xxs', 'dress'],
    ['xxs', 'skirt'],
    ['xs', 'dress'],
    ['s', 'skirt'],
    ['m', 'dress'],
    ['l', 'shirt'],
    ['s', 'coat'],
    ['m', 'coat'],
    ['xxl', 'shirt'],
    ['l', 'dress']
]
clothing = pd.DataFrame(clothing_list, columns=['size', 'type'])
display(clothing)

size   type
0  xxs  dress
1  xxs  skirt
2   xs  dress
3    s  skirt
4    m  dress
5    l  shirt
6    s   coat
7    m   coat
8  xxl  shirt
9    l  dress

In [37]:
ord_encoder = ce.OrdinalEncoder()
data_bin = ord_encoder.fit_transform(clothing[['size', 'type']])
clothing = pd.concat([clothing, data_bin], axis=1)
display(clothing)

size   type  size  type
0   xxs  dress     1     1
1   xxs  skirt     1     2
2    xs  dress     2     1
3     s  skirt     3     2
4     m  dress     4     1
5     l  shirt     5     3
6     s   coat     3     4
7     m   coat     4     4
8   xxl  shirt     6     3
9     l  dress     5     1

###  Задание 5.3 (Самопроверка) ###

Использовать ранее изученные методы кодирования и закодировать признак `year` в датасете винных обзоров порядковым кодированием.

In [39]:
data_bin = ord_encoder.fit_transform(df[['year']])

## Однократное кодирование. One-Hot Encoding ##

In [42]:
encoder = ce.OneHotEncoder(cols=['type'])
type_bin = encoder.fit_transform(clothing['type'])
clothing = pd.concat([clothing, type_bin], axis=1)
display(clothing)

AttributeError: 'DataFrame' object has no attribute 'unique'